File: number_pub_and_counter_py/number_pub_and_counter_py/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
    
    
class NumberCounter(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.logger = self.get_logger()
        self.counter = 0
        self.count_publishers_ = self.create_publisher(Int64, "number_count", 10)
        self.num_subscriber_ = self.create_subscription(Int64, "number", self.numPubCallback, 10)
        self.reset_counter_server_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
        self.logger.info("Number counter node has started")

    def numPubCallback(self, msg: Int64):
        self.counter += 1
        self.logger.info(f"Number received: {msg.data} {self.counter} times")
        new_msg = Int64()
        new_msg.data = self.counter
        self.count_publishers_.publish(msg=new_msg)
    
    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data:
            self.get_logger().info(f"Resetting counter from {self.counter} to 0")
            response.success = True
            response.message = f"Resetting counter from {self.counter} to 0"
            self.counter = 0
        else:
            response.message = f"Data for reuqest is {request.data}, keeping counter at {self.counter}"
        return response
    
    
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounter()
    rclpy.spin(node)
    rclpy.shutdown()
    
    
if __name__ == "__main__":
    main()

File: number_pub_and_counter_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
    
class NumberCounter : public rclcpp::Node
{
public:
    NumberCounter() : Node("node_name")
    {
        counter_ = 0;
        count_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        num_subscriber_ = this->create_subscription<example_interfaces::msg::Int64>("number", 10,
                                std::bind(&NumberCounter::numCallback, this, std::placeholders::_1));
    }
    
private:
    void numCallback(example_interfaces::msg::Int64 msg){
        counter_++;
        RCLCPP_INFO(this->get_logger(), "%ld number received %d times", msg.data, counter_);
        auto new_msg = example_interfaces::msg::Int64();
        new_msg.data = counter_;
        count_publisher_->publish(new_msg);
    }

    int counter_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr count_publisher_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr num_subscriber_;
};
    
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}